netket/optimizer/jax_stochastic_reconfiguration.py

```diff
@@ -0,0 +1,370 @@
+
+        self._comm = MPI.COMM_WORLD
+
+    def _init_solver(self):
+        lsq_solver = self._lsq_solver
+
+        if lsq_solver in ["gmres", "cg", "minres","jaxcg"]:
+            self._use_iterative = True
+        if lsq_solver in ["ColPivHouseholder", "QR", "SVD", "Cholesky"]:
+            self._use_iterative = False
+
+        if self._use_iterative:
+            if lsq_solver is None:
+                lsq_solver = "jaxcg"
+
+            if lsq_solver == "gmres":
+                self._sparse_solver = partial(gmres, atol="legacy")
+            elif lsq_solver == "cg":
+                self._sparse_solver = partial(cg, atol="legacy")
+            elif lsq_solver == "minres":
+                if self._is_holomorphic:
+                    raise RuntimeError(
+                        "minres can be used only for real-valued parameters."
+                    )
+                self._sparse_solver = minres
+            elif not lsq_solver == 'jaxcg':
+                raise RuntimeError("Unknown sparse lsq_solver " + lsq_solver + ".")
+
+        else:
+            if (
+                lsq_solver is None
+                or "ColPivHouseholder" in lsq_solver
+                or "QR" in lsq_solver
+            ):
+                self._lapack_driver = "gelsy"
+            elif "SVD" in lsq_solver or "Cholesky" in lsq_solver:
+                self._lapack_driver = None
+            else:
+                self._lapack_driver = None
+                raise RuntimeError("Unknown lsq_solver" + lsq_solver + ".")
+
+        if self._use_iterative and self._svd_threshold is not None:
+            raise ValueError(
+                "The svd_threshold option is available only for non-sparse solvers."
+            )
+
+    def compute_update(self, oks, grad, out=None):
+        r"""
+        Solves the SR flow equation for the parameter update ẋ.
+
+        The SR update is computed by solving the linear equation
+           Sẋ = f
+        where S is the covariance matrix of the partial derivatives
+        O_i(v_j) = ∂/∂x_i log Ψ(v_j) and f is a generalized force (the loss
+        gradient).
+
+        Args:
+            oks: The matrix of log-derivatives,
+		O_i(v_j)
+            grad: The vector of forces f.
+            out: Output array for the update ẋ.
+        """
+
+        oks -= jax.numpy.mean(oks,axis=0)
+
+        if self.is_holomorphic is None:
+            raise ValueError(
+                "is_holomorphic not set: this SR object is not properly initialized."
+            )
+
+        n_samp = _sum_inplace(_np.atleast_1d(oks.shape[0]))
+
+        n_par = grad.shape[0]
+
+        if out is None:
+            out = _np.zeros(n_par, dtype=_np.complex128)
+
+        if self._is_holomorphic:
+            if self._use_iterative:
+                if self._lsq_solver == "jaxcg":
+                    out = self._jax_cg_solve(oks,grad,n_samp)
+                else:
+                    op = self._linear_operator(oks, n_samp)
```

Comment:
I think we can remove this branch that doesn't use jax solvers

Reply:
Okay, do we want to keep the scipy exact solvers though? Right now the Jax ones don't appear to be working

Reply:
is there a bug in their non-sparse solvers? Can you reproduce it with a simple example?

Reply:
Yes its a bug with their non-sparse solvers. As an example for cho_factor, this script returns a bus error

import numpy as np
import jax.numpy as jnp
from jax.scipy.linalg import cho_factor

vec = jnp.array(np.random.normal(0.,0.1,[20,400]) + 1j*np.random.normal(0.,0.1,[20,400]))

matrix = jnp.matmul(vec.transpose(),vec) + jnp.array(np.identity(400))

print(cho_factor(matrix))

Reply:
The threshold for breaking is slightly different for all of the solvers but a 400x400 matrix breaks all of them so they're useless at the moment

Reply:
But this specific case breaks also on scipy, as far as I can tell 
```
 raise LinAlgError("%d-th leading minor of the array is not positive "
numpy.linalg.LinAlgError: 29-th leading minor of the array is not positive definite
```

are you sure there isn't something else going wrong?

Reply:
I mean the issue is that you didn't put the conjugate, so that's not even a positive-definite matrix

```
matrix = jnp.matmul(vec.transpose().conjugate(),vec) + jnp.array(np.identity(400))
```
this works

Reply:
Sorry my mistake. This code breaks in jax but not numpy

import numpy as np
import jax.numpy as jnp
from jax.scipy.linalg import cho_factor

vec = jnp.array(np.random.normal(0.,0.1,[20,400]) + 1j*np.random.normal(0.,0.1,[20,400]))

matrix = jnp.matmul(vec.transpose().conjugate(),vec) + jnp.array(np.identity(400))

print(cho_factor(matrix))

Reply:
Numpy code that works:

import numpy as np
from scipy.linalg import cho_solve, cho_factor

vec = np.random.normal(0.,0.1,[20,400]) + 1j*np.random.normal(0.,0.1,[20,400])

matrix = np.matmul(vec.transpose().conjugate(),vec) + np.identity(400)

print(cho_factor(matrix))

Reply:
i had forgotten the conjugate

Reply:
This works fine on my computer
```
import numpy as np
import jax.numpy as jnp
from jax.scipy.linalg import cho_factor

n=1000

vec = jnp.array(np.random.normal(0.,0.1,[20,n]) + 1j*np.random.normal(0.,0.1,[20,n]))

matrix = jnp.matmul(vec.transpose().conjugate(),vec) + jnp.array(np.identity(n))

print(cho_factor(matrix))
```

Reply:
Yes, that script also breaks on my computer

Reply:
Yes, I confirm that everything seems to work fine on my computer, also the solve function

```python
import numpy as np
import jax.numpy as jnp
from jax.scipy.linalg import cho_factor,cho_solve

n=1000

vec = jnp.array(np.random.normal(0.,0.1,[20,n]) + 1j*np.random.normal(0.,0.1,[20,n]))
b=jnp.array(np.random.normal(size=n)+1.0j*np.random.normal(size=n))
matrix = jnp.matmul(vec.transpose().conjugate(),vec) + jnp.array(np.identity(n))

fac=cho_factor(matrix)
sol=cho_solve(fac,np.asarray(b))

vec1=np.asarray(vec)
matrix1=np.asarray(matrix)


import scipy
from scipy.linalg import cho_factor
fac=cho_factor(matrix1)

sol1=scipy.linalg.cho_solve(fac,np.asarray(b))

assert(np.linalg.norm(sol-sol1,ord=np.inf)<1.0e-5)
```

Reply:
I have a mac, I remember having similar bizarre issues with pytorch as well, so maybe i'll just implement and test with small arrays

Reply:
import numpy as np
import jax.numpy as jnp
from jax.scipy.linalg import cho_factor,cho_solve,qr,solve_triangular,svd

n=30

vec = jnp.array(np.random.normal(0.,0.1,[20,n]) + 1j*np.random.normal(0.,0.1,[20,n]))
b=jnp.array(np.random.normal(size=n)+1.0j*np.random.normal(size=n))
matrix = jnp.matmul(vec.transpose().conjugate(),vec) + jnp.array(np.identity(n))

#cho
fac=cho_factor(matrix)
sol=cho_solve(fac,np.asarray(b))
assert(np.linalg.norm(jnp.matmul(matrix,sol)-b,ord=np.inf)<1.0e-5)

#qr
Q, R = qr(matrix)
c = jnp.matmul(Q.transpose().conjugate(),b)
sol = solve_triangular(R,c)
assert(np.linalg.norm(jnp.matmul(matrix,sol)-b,ord=np.inf)<1.0e-5)

#SVD
U, S, V = svd(matrix)
c = jnp.matmul(U.transpose().conjugate(),b)/S
sol = jnp.matmul(V.transpose().conjugate(),c)
assert(np.linalg.norm(jnp.matmul(matrix,sol)-b,ord=np.inf)<1.0e-5)

Here's the code I used to check the Jax solvers were working correctly. Unfortunately this breaks for larger "n" so it would be nice if someone gave it a shot on their computers.

Reply:
It works fine (I tried up to n=1000). For n=1000 you might need to use double precision to satisfy the assertion though
```python
import os
os.environ["JAX_ENABLE_X64"] = "1"

import numpy as np
import jax.numpy as jnp
from jax.scipy.linalg import cho_factor,cho_solve,qr,solve_triangular,svd
```

Reply:
So most likely it's an issue with your installation (I have mac too, and I am using this on conda)